hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java

```diff
@@ -130,6 +131,13 @@ public boolean cancel(boolean mayInterruptIfRunning) {
                 return false;
             }
 
+            if (shouldCancel == null) {
+                shouldCancel = shouldCancel(mayInterruptIfRunning);
+                if (!shouldCancel) {
+                    return false;
+                }
+            }
+
```

Comment:
For me this would be clearer if the local variable was `alreadyAskedShouldCancel` or similar because you only check the actual `shouldCancel` result once. Currently the semantics of the tri-state variable are confusing.